genesis/utils/path_planing.py

```diff
@@ -0,0 +1,995 @@
+                    envs_idx=envs_idx,
+                    is_plan_with_obj=is_plan_with_obj,
+                    obj_geom_start=obj_geom_start,
+                    obj_geom_end=obj_geom_end,
+                    ee_link_idx=ee_link_idx,
+                    obj_link_idx=obj_link_idx,
+                    T_robot_object=T_robot_obj,
+                )
+            else:
+                sol = self.shortcut_path(
+                    torch.ones_like(sol[..., 0]), sol, iterations=10, ignore_geom_pairs=unique_pairs, envs_idx=envs_idx
+                )
+
+        sol = align_weypoints_length(sol, torch.ones_like(sol[..., 0]).bool(), num_waypoints)
+
+        if ignore_collision:
+            is_invalid = is_active
+        else:
+            if is_plan_with_obj:
+                is_invalid = torch.logical_or(
+                    is_active,
+                    self.check_collision(
+                        sol,
+                        unique_pairs,
+                        envs_idx,
+                        is_plan_with_obj=is_plan_with_obj,
+                        obj_geom_start=obj_geom_start,
+                        obj_geom_end=obj_geom_end,
+                        ee_link_idx=ee_link_idx,
+                        obj_link_idx=obj_link_idx,
+                        T_robot_object=T_robot_obj,
+                    ),
+                )
+            else:
+                is_invalid = torch.logical_or(
+                    is_active,
+                    self.check_collision(
+                        sol,
+                        unique_pairs,
+                        envs_idx,
+                    ),
+                )
+
+        if self._solver.n_envs > 0:
+            self._entity.set_qpos(qpos_cur, envs_idx=envs_idx)
+        else:
+            self._entity.set_qpos(qpos_cur)
+
+        if is_plan_with_obj:
+            self.update_object(ee_link_idx, obj_link_idx, T_robot_obj, envs_idx)
+
+        if is_invalid.any():
+            num_failure = int(is_invalid.sum().cpu())
+            gs.logger.warning(f"rrt connect planning failed in {num_failure} environments")
+        return sol, is_invalid
+
+
+# ------------------------------------------------------------------------------------
+# ------------------------------------ utils -----------------------------------------
+# ------------------------------------------------------------------------------------
+
+
+def align_weypoints_length(
+    path: torch.Tensor, mask: torch.Tensor, num_points: int
+) -> torch.Tensor:
+    """
+    Aligns each waypoints length to the given num_points.
+
+    Parameters
+    ----------
+    path: torch.Tensor
+        path tensor in [N, B, Dof]
+    mask: torch.Tensor
+        the masking of path, indicating active waypoints [N, B]
+    num_points: int
+        the number of the desired waypoints
+
+    Returns
+    -------
+        A new 2D PyTorch tensor [num_points, B, Dof]
+    """
+    t_path = path.permute(1, 2, 0) # [B, Dof, N]
```

Comment:
What about doing `t_path = path.permute(1, 0, 2)`? This way you should be able to remove transpose in
`res[:, i_b] = interpolated_path`? Not a big deal though. It is already readable enough.